### ISSUE TITLE
reference build failure

### ISSUE THREAD

**Issue by ChiaraCaprioli** [williamritchie/IRFinder#168]
Hi, 

I am trying to build the IRFinder reference by running the following (files have been downloaded directly from Ensembl):
```
bin/IRFinder -m BuildRefProcess -r REF/Human-GRCh38-release98
```

However, this is what happens:
```
Launching reference build process. The full build might take hours.
<Phase 1: STAR Reference Preparation>
PetaLink: Can't intercept static program '/hpcnfs/home/ieo4874/.conda/envs/IRFinder/bin/STAR' without +static
	STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/.conda/envs/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/.conda/envs/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/.conda/envs/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28 &>> log-star-build-ref.log
	STAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Feb 13 18:45:39 ..... started STAR run
Feb 13 18:45:39 ... starting to generate Genome files
Feb 13 18:46:31 ..... processing annotations GTF
Feb 13 18:47:04 ... starting to sort Suffix Array. This may take a long time...
Feb 13 18:47:21 ... sorting Suffix Array chunks and saving them to disk...
Star genome build: Failed!
```

These are my settings:

IRFinder=1.3.0
bedtools=2.30.0              
gcc=12.2.0
perl=5.32.1 
samtools=1.6 
star=2.7.10b

```
cat /etc/*-release
```
CentOS Linux release 7.9.2009 (Core)
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

CentOS Linux release 7.9.2009 (Core)
CentOS Linux release 7.9.2009 (Core)

```
ldd --version
```
ldd (GNU libc) 2.17

Is there anything else I can double-check?
Thank you

Chiara

**Comment by dg520** [comment]
@ChiaraCaprioli Did you install IRFinder via conda? I am not the maintainer of the conda-version of IRFinder, so unfortunately I have very limited insight into that package. It looks like `STAR` is not complied in a functional way. It seems to look for a static library but cannot find it. 
Do you have another version of `STAR` that is directly callable (i.e. a callable STAR path included in your Linux environmental variable `$PATH` )? If so, I would recommend downloading and installing IRFinder directly from GitHub.

**Comment by ChiaraCaprioli** [comment]
I installed IRFinder directly from Github.
The issue with the static library seems solved by disabling Petalink (at least, the error disappeared). However the rest of the error looks more or less the same as before:
```
Launching reference build process. The full build might take hours.
<Phase 1: STAR Reference Preparation>
	STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28 &>> log-star-build-ref.log
	STAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Feb 16 16:07:39 ..... started STAR run
Feb 16 16:07:39 ... starting to generate Genome files
Star genome build: Failed!
```
Following you find the ```Log.out``` from ```REF/Human-GRCh38-release98/STAR```:
```
STAR version=2.7.10b
STAR compilation time,server,dir=2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
STAR git: On branch master ; commit c6f8efc2c7043ef83bf8b0d9bed36bbb6b9b1133 ; diff files: CHANGES.md 
##### Command Line:
STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28 &>> log-star-build-ref.log
##### Initial USER parameters from Command Line:
###### All USER parameters from Command Line:
runMode                       genomeGenerate        ~RE-DEFINED
genomeDir                     /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR     ~RE-DEFINED
genomeFastaFiles              /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa        ~RE-DEFINED
sjdbGTFfile                   /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf     ~RE-DEFINED
sjdbOverhang                  150     ~RE-DEFINED
runThreadN                    28     ~RE-DEFINED
##### Finished reading parameters from all sources

##### Final user re-defined parameters-----------------:
runMode                           genomeGenerate   
runThreadN                        28
genomeDir                         /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR
genomeFastaFiles                  /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa   
sjdbGTFfile                       /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf
sjdbOverhang                      150

-------------------------------
##### Final effective command line:
STAR   --runMode genomeGenerate      --runThreadN 28   --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR   --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa      --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf   --sjdbOverhang 150
----------------------------------------

Number of fastq files for each mate = 1
ParametersSolo: --soloCellFilterType CellRanger2.2 filtering parameters:  3000 0.99 10
Finished loading and checking parameters
--genomeDir directory exists and will be overwritten: /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR/
Feb 16 16:07:39 ... starting to generate Genome files
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 0  "1" chrStart: 0
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 1  "10" chrStart: 249036800
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 2  "11" chrStart: 382992384
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 3  "12" chrStart: 518258688
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 4  "13" chrStart: 651689984
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 5  "14" chrStart: 766246912
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 6  "15" chrStart: 873463808
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 7  "16" chrStart: 975699968
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 8  "17" chrStart: 1066139648
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 9  "18" chrStart: 1149501440
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 10  "19" chrStart: 1229979648
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 11  "2" chrStart: 1288699904
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 12  "20" chrStart: 1530920960
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 13  "21" chrStart: 1595408384
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 14  "22" chrStart: 1642332160
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 15  "3" chrStart: 1693188096
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 16  "4" chrStart: 1891631104
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 17  "5" chrStart: 2081947648
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 18  "6" chrStart: 2263613440
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 19  "7" chrStart: 2434531328
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 20  "8" chrStart: 2593914880
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 21  "9" chrStart: 2739142656
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 22  "MT" chrStart: 2877554688
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 23  "X" chrStart: 2877816832
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa : chr # 24  "Y" chrStart: 3034054656
```

**Comment by dg520** [comment]
@ChiaraCaprioli To be honest, this looks like a `STAR` issue instead of `IRFinder`. I can try to provide some insights with some further information from you. But I cannot guarantee I can solve it. You might still need to ask the `STAR` team for the solution. But let's try first. Can you find `log-star-build-ref.log` and share me the content, or is that what you posted above already although you said it is from `Log.out`?

**Comment by ChiaraCaprioli** [comment]
It seems to have worked

```
cat log-star-build-ref.log
STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28
	STAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Feb 16 20:52:54 ..... started STAR run
Feb 16 20:52:54 ... starting to generate Genome files
	STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28
	STAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Feb 16 20:53:41 ..... started STAR run
Feb 16 20:53:41 ... starting to generate Genome files
	STAR --runMode genomeGenerate --genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR --genomeFastaFiles /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa --sjdbGTFfile /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf --sjdbOverhang 150 --runThreadN 28
	STAR version: 2.7.10b   compiled: 2022-11-01T09:53:26-04:00 :/home/dobin/data/STAR/STARcode/STAR.master/source
Feb 16 20:54:29 ..... started STAR run
Feb 16 20:54:29 ... starting to generate Genome files
Feb 16 20:55:24 ..... processing annotations GTF
Feb 16 20:55:57 ... starting to sort Suffix Array. This may take a long time...
Feb 16 20:56:14 ... sorting Suffix Array chunks and saving them to disk...
```

**Comment by dg520** [comment]
@ChiaraCaprioli 
OK, great. I think your previous problem is that your STAR prints out some unusual messages in standard errors and IRFinder thinks STAR has failed. But those messages are NOT errors and should not be printed as standard errors in the first place. I'm not sure why this happens to you but most likely something funky there during STAR compilation. 
Nevertheless, you can go ahead to complete IRFinder reference building by running the following line:
```
IRFinder -m BuildRefFromSTARRef \
-r /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder \
-x /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR \
-e /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/extra-input-files/RNA.SpikeIn.ERCC.fasta.gz \
-R /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/extra-input-files/Human_hg38_nonPolyA_ROI.bed
```
This will generate the reference at IRFinder reference at `/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder`

**Comment by ChiaraCaprioli** [comment]
Hi, I'm sorry to bother again, but even though the process can be completed, the reference files I obtained are empty:

```
Launching reference build process. The full build might take hours.
<Phase 1: STAR Reference Preparation>
Feb 17 12:57:40 ... copying the genome FASTA file...
Feb 17 12:58:01 ... copying the transcriptome GTF file...
Feb 17 12:58:12 ... copying the STAR reference folder...
<Phase 2: Mapability Calculation>
Feb 17 13:01:29 ... mapping genome fragments back to genome...
/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/bin/util/Mapability: line 36: 11562 Killed                  "$STAREXEC" --genomeDir "$STARGENOME" --genomeLoad NoSharedMemory --runThreadN $THREADS --outStd SAM --outSAMmode NoQS --outSAMattributes None --outFilterMultimapNmax 1 --readFilesIn <("$LIBEXEC/generateReadsError.pl" 70 10 < "$FA") > genome_fragments.sam
Feb 17 13:02:27 ... sorting aligned genome fragments...
Feb 17 13:02:27 ... indexing aligned genome fragments...
Feb 17 13:02:27 ... filtering aligned genome fragments by chromosome/scaffold...
[main_samview] region "1" specifies an unknown reference name. Continue anyway.
[main_samview] region "12" specifies an unknown reference name. Continue anyway.
[main_samview] region "11" specifies an unknown reference name. Continue anyway.
[main_samview] region "14" specifies an unknown reference name. Continue anyway.
[main_samview] region "10" specifies an unknown reference name. Continue anyway.
[main_samview] region "15" specifies an unknown reference name. Continue anyway.
[main_samview] region "13" specifies an unknown reference name. Continue anyway.
[main_samview] region "17" specifies an unknown reference name. Continue anyway.
[main_samview] region "16" specifies an unknown reference name. Continue anyway.
[main_samview] region "20" specifies an unknown reference name. Continue anyway.
[main_samview] region "18" specifies an unknown reference name. Continue anyway.
[main_samview] region "21" specifies an unknown reference name. Continue anyway.
[main_samview] region "4" specifies an unknown reference name. Continue anyway.
[main_samview] region "22" specifies an unknown reference name. Continue anyway.
[main_samview] region "6" specifies an unknown reference name. Continue anyway.
[main_samview] region "3" specifies an unknown reference name. Continue anyway.
[main_samview] region "19" specifies an unknown reference name. Continue anyway.
[main_samview] region "2" specifies an unknown reference name. Continue anyway.
[main_samview] region "MT" specifies an unknown reference name. Continue anyway.
[main_samview] region "Y" specifies an unknown reference name. Continue anyway.
[main_samview] region "7" specifies an unknown reference name. Continue anyway.
[main_samview] region "8" specifies an unknown reference name. Continue anyway.
[main_samview] region "5" specifies an unknown reference name. Continue anyway.
[main_samview] region "9" specifies an unknown reference name. Continue anyway.
[main_samview] region "X" specifies an unknown reference name. Continue anyway.
Feb 17 13:02:27 ... merging filtered genome fragments...

gzip: stdin: unexpected end of file
Feb 17 13:02:27 ... calculating regions for exclusion...
Feb 17 13:02:41 ... cleaning temporary files...
rm: missing operand
Try 'rm --help' for more information.
rm: cannot remove 'SJ.out.tab': No such file or directory
<Phase 3: IRFinder Reference Preparation>
Feb 17 13:02:41 ... building Ref 1...
Feb 17 13:03:17 ... building Ref 2...
Feb 17 13:03:17 ... building Ref 3...
Feb 17 13:03:17 ... building Ref 4...
Feb 17 13:03:25 ... building Ref 5...
Feb 17 13:03:37 ... building Ref 6...
Feb 17 13:03:37 ... building Ref 7...
Feb 17 13:03:37 ... building Ref 8...
Feb 17 13:03:38 ... building Ref 9...
Feb 17 13:03:38 ... building Ref 10c...
Feb 17 13:03:38 ... building Ref 11c...
Error: ref-cover.bed is empty.
Error: ref-read-continues.ref is empty.
Error: IRFinder reference building FAILED.
```

I checked your troubleshooting suggestions and run ```ulimit -n``` accordingly, which results to 16384. Other than this I don't know what else I could check

**Comment by dg520** [comment]
@ChiaraCaprioli Interesting, it still failed at a pure `STAR` step. I really think your STAR has not been compiled correctly. Can we have one more test before I run out of ideas? Try the following commands:
```
mkdir -p /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/Mapability;
cd /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/Mapability;
(/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/bin/util/generateReadsError.pl 70 10 < /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/genome.fa) > tmp.fq
```
Does it give you error? If so, tell me what error it is. If not, continue with the following:
```
STAR \
--genomeDir /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR \
--genomeLoad NoSharedMemory \
--runThreadN 28 \
--outStd SAM \
--outSAMmode NoQS \
--outSAMattributes None \
--outFilterMultimapNmax 1 \
--readFilesIn tmp.fq > genome_fragments.sam
```
Does it work or fail?

**Comment by ChiaraCaprioli** [comment]
It worked, I could obtain the ```genome_fragments.sam``` file. 
Thanks a lot

**Comment by dg520** [comment]
Did it? That is interesting. Not sure why it is successful when you run it manually. 
Now you have to complete the remaining steps of IRFinder reference preparation manually as well. Try the following:
```
cd /hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/Mapability;
LIBEXEC=/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/bin/util
STARGENOME=/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/STAR
THREADS=28
TMPBED=tmp_$$
TMPCMP=gzip
TMPEXT=gz

mkdir "$TMPBED"
samtools sort -@ "$THREADS" genome_fragments.sam > genome_fragments.bam
samtools index -@ "$THREADS" genome_fragments.bam

cat "$STARGENOME/chrName.txt" | \
    xargs --max-args 1 --max-procs "$THREADS" -I{} bash -c "samtools view genome_fragments.bam {}|awk -v tmpdir=\"$TMPBED\" -v tmpcmp=\"$TMPCMP\" -v tmpext=\"$TMPEXT\" 'BEGIN{FS=\"[\\t!]\"; OFS=\"\\t\"}{if ((\$8 == \"70M\") && (\$3 == \$6) && (\$2 == \$5)) {print \$5, \$6-1, \$6+69 | (tmpcmp \" -c1 > \" tmpdir \"/\" \$5 \".bed.\" tmpext ) }}END{close( (tmpcmp \" -c1 > \" tmpdir \"/\" \$5 \".bed.\" tmpext ))}'"

find "$TMPBED" -type f -name "*.bed.""$TMPEXT"|xargs --max-args 1 zcat >> genome_fragments.unsorted.bed

bedtools genomecov -i genome_fragments.unsorted.bed -bga -g "$STARGENOME/chrNameLength.txt" | \
    awk 'BEGIN{FS=OFS="\t";chr="random"}($1!=chr){chr=$1}($1==chr){print}' | \
    awk 'BEGIN {FS=OFS="\t"} ($4 < 5) {print $1, $2, $3}' | \
    bedtools merge -i stdin | \
    sort -S5G -k1,1 -k2,2n -k3,3n| \
    gzip > MapabilityExclusion.bed.gz

find "$TMPBED" -type f -name "*.bed.""$TMPEXT"|xargs --max-args 1 --max-procs "$THREADS" rm
rm genome_fragments.*
rm Log.*
rm SJ.out.tab
rmdir "$TMPBED"
```
and then 
```
basedir=/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/

"$LIBEXEC/Build-BED-refs.sh" \
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/transcripts.gtf \
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/STAR/chrNameLength.txt \
"$basedir/Mapability/MapabilityExclusion.bed.gz" \
$LIBEXEC \
/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/extra-input-files/Human_hg38_nonPolyA_ROI.bed \
""
```

**Comment by ChiaraCaprioli** [comment]
Hi Dadi,

your last solution worked fine, thank you.
However...

I tried to go on with IR quantification in the BAM mode, using the reference obtained with the above steps and a sorted bam
```
bin/IRFinder -m BAM -r REF/Human-GRCh38-release98 /hpcnfs/scratch/PGP/ccaprioli/working/EIF4A3/data/bam/MDA-MB-468/12998.bam -d /hpcnfs/scratch/PGP/ccaprioli/working/EIF4A3/data/IRFinder/12998
```

and I got an error containing
```
... version `GLIBCXX_xxx' not found ....
```

According to the troubleshooting section, I ran
```
cd src/irfinder
make clean; make
```

but compilation fails, as follows 
```
Makefile:32: Depend.list: No such file or directory
/bin/rm -f ./Depend.list
g++ -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   -MM ReadBlockProcessor.cpp FragmentBlocks.cpp IRFinder.cpp crc32.cpp ReadBlockProcessor_OutputBAM.cpp CoverageBlock.cpp ReadBlockProcessor_CoverageBlocks.cpp BAM2blocks.cpp >> Depend.list
make: Warning: File `Depend.list' has modification time 0.014 s in the future
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   FragmentBlocks.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   ReadBlockProcessor.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   CoverageBlock.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   ReadBlockProcessor_CoverageBlocks.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   ReadBlockProcessor_OutputBAM.cpp
ReadBlockProcessor_OutputBAM.cpp:184:6: warning: unused parameter ‘chrmap’ [-Wunused-parameter]
 void OutputBAM::ChrMapUpdate(const std::vector<string> &chrmap) {
      ^
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   BAM2blocks.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   IRFinder.cpp
g++ -c -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   crc32.cpp
g++ -o irfinder -pipe -std=c++0x -O3    -Wall -Wextra -fopenmp -D'COMPILATION_TIME_PLACE="Sat Feb 18 17:58:27 CET 2023 cn07.cluster.loc:/hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/src/irfinder"'   FragmentBlocks.o ReadBlockProcessor.o CoverageBlock.o ReadBlockProcessor_CoverageBlocks.o ReadBlockProcessor_OutputBAM.o BAM2blocks.o IRFinder.o crc32.o -static -static-libgcc
/bin/ld: cannot find -lstdc++
/bin/ld: cannot find -lm
/bin/ld: cannot find -lpthread
/bin/ld: cannot find -lc
collect2: error: ld returned 1 exit status
make: *** [IRFinderStatic] Error 1
```

Both gcc and glibc seem updated.
```
gcc --version
gcc (conda-forge gcc 12.2.0-19) 12.2.0
```
```
ldd --version
ldd (GNU libc) 2.17
```

**Comment by dg520** [comment]
@ChiaraCaprioli IRFinder cannot find multiple basic C/C++ libraries at the default location of your system, including `libstdc++`, `libm`, `libpthread` and `libc`. Have a read [this](https://github.com/williamritchie/IRFinder/wiki/Download-and-Install), which is an example of how to solve the unfound `libm` issue that is more common. You can do the same to locate the other missing libraries. 

If you're uncomfortable of doing so, you can try to use your conda-version of IRFinder. Previously it failed to prepare the reference due to the mis-compilation of STAR. Now you've a reference in hand (hopefully), you can try to call the IR quantification from there. Maybe its IRFinder core function has been correctly compiled and that is very likely. 

Otherwise, you have to contact your system admin and ask them to install and compile IRFinder for you from GitHub.

**Comment by ChiaraCaprioli** [comment]
Hi, 

with the conda version of IRFinder seems working. However, I get the following warning:

```This sample has excessive splice junctions at unannotated locations. This may indicate the experiment is not actually RNA-Seq. Or it indicates the genome fasta and annotation gtf were not compatible.```

As I found a different chromosome annotation in the built reference and my test bam file, I changed the annotation in the genome.fa and transcripts.gtf, but got the same error. Then I tried again changing the bam as follows

```
samtools view -h sorted_12998.bam | \
sed -e '/^@SQ/s/SN\:chr/SN\:/' -e '/^[^@]/s/\tchr/\t/2' | \
samtools view -bS - > temp.bam
```

but another comes out, namely:
```
/hpcnfs/home/ieo4874/.conda/envs/IRFinder2/bin/IRFinder -m BAM \
-r /hpcnfs/data/PGP/ccaprioli/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98 \
/hpcnfs/scratch/PGP/ccaprioli/working/EIF4A3/data/bam/MDA-MB-468/temp.bam 
Illegal division by zero at /hpcnfs/home/ieo4874/.conda/envs/IRFinder2/opt/irfinder-1.3.1/bin/util/warnings line 207.
```

Is there anything else I could check/change?
Thank you

**Comment by dg520** [comment]
@ChiaraCaprioli  
Choose one of the following approach:
1) keep your original BAM unchanged. Change the chromosome names in BOTH your `genome.fa` and `transcripts.gtf` according to your BAM. And remake the IRFinder reference. Quantify IR using BAM mode against the NEW IRFinder reference.
or 
2) keep your original `genome.fa`, `transcripts.gtf` and IRFinder reference unchanged. Regenerate the BAM file against the STAR reference at  `/hpcnfs/home/ieo4874/tools/IRFinder/IRFinder-1.3.0/REF/Human-GRCh38-release98/IRFinder/STAR`. Quantify IR using BAM mode with the NEW BAM file.
or 
3) Nothing has to be changed. Just quantify IR using FASTQ mode, instead of BAM mode, using FASTQ files. 

Best,
Dadi

**Comment by ChiaraCaprioli** [comment]
Thanks a lot, it's solved by approach #3.
